src/tools/cephfs/DataScan.h

```diff
@@ -0,0 +1,230 @@
+     */
+    virtual int inject_lost_and_found(
+        inodeno_t ino,
+        uint64_t size,
+        time_t mtime,
+        uint32_t chunk_size,
+        int64_t data_pool_id) = 0;
+
+    /**
+     * Create any missing roots (i.e. mydir, strays, root inode)
+     */
+    virtual int init_roots(
+        int64_t data_pool_id) = 0;
+
+    /**
+     * Pre-injection check that all the roots are present in
+     * the metadata pool.  Used to avoid parallel workers interfering
+     * with one another, by cueing the user to go run 'init' on a
+     * single node before running a parallel scan.
+     *
+     * @param result: set to true if roots are present, else set to false
+     * @returns 0 on no unexpected errors, else error code.  Missing objects
+     *          are not considered an unexpected error: check *result for
+     *          this case.
+     */
+    virtual int check_roots(bool *result) = 0;
+
+    /**
+     * Helper to compose dnames for links to lost+found
+     * inodes.
+     */
+    std::string lost_found_dname(inodeno_t ino)
+    {
+      char s[20];
+      snprintf(s, sizeof(s), "%llx", (unsigned long long)ino);
+      return std::string(s);
+    }
+
+    virtual ~RecoveryDriver() {}
+};
+
+class LocalFileDriver : public RecoveryDriver
+{ 
+  protected:
+    const std::string path;
+    librados::IoCtx &data_io;
+
+  int inject_data(
+      const std::string &file_path,
+      uint64_t size,
+      uint32_t chunk_size,
+      inodeno_t ino);
+  public:
+
+    LocalFileDriver(const std::string &path_, librados::IoCtx &data_io_)
+      : path(path_), data_io(data_io_)
+    {}
+
+    // Implement RecoveryDriver interface
+    int init(librados::Rados &rados, const MDSMap *mdsmap);
+
+    int inject_with_backtrace(
+        const inode_backtrace_t &bt,
+        uint64_t size,
+        time_t mtime,
+        uint32_t chunk_size,
+        int64_t data_pool_id);
+
+    int inject_lost_and_found(
+        inodeno_t ino,
+        uint64_t size,
+        time_t mtime,
+        uint32_t chunk_size,
+        int64_t data_pool_id);
+
+    int init_roots(int64_t data_pool_id);
+
+    int check_roots(bool *result);
+};
+
+/**
+ * A class that knows how to manipulate CephFS metadata pools
+ */
+class MetadataDriver : public RecoveryDriver
+{
+  protected:
+
+    librados::IoCtx metadata_io;
+
+    /**
+     * Create a .inode object, i.e. root or mydir
+     */
+    int inject_unlinked_inode(inodeno_t inono, int mode, int64_t data_pool_id);
+
+    /**
+     * Check for existence of .inode objects, before
+     * trying to go ahead and inject metadata.
+     */
+    int root_exists(inodeno_t ino, bool *result);
+
+    /**
+     * Try and read an fnode from a dirfrag
+     */
+    int read_fnode(inodeno_t ino, frag_t frag, fnode_t *fnode);
+
+    /**
+     * Try and read a dentry from a dirfrag
+     */
+    int read_dentry(inodeno_t parent_ino, frag_t frag,
+        const std::string &dname, InodeStore *inode);
```

Comment:
Did you change your editor or something? This function started out nicely indented (with parameters aligned) and then got "busted". I don't know if we actually have standards on the tabbing besides what emacs does with the command lines we have in our old files, but it's throwing me for a loop. ;)